### ISSUE TITLE
Cannot add a srcDir including precompiled script plugins

### ISSUE THREAD

**Issue by hakanai** [gradle/gradle#21052]
### Expected Behavior

It should be possible to add a `srcDir("src/main/kotlin2")` or wherever to add additional source directories. If precompiled script plugins exist in this directory as well, they should be treated the same as if they were in `src/main/kotlin`.

### Current Behavior

Currently, `src/main/kotlin` seems to be treated specially - the compiler for those is somehow configured differently to other `srcDirs` in the project. This causes them to fail compilation on basic things such as the `allprojects {}` block.

### Context

The thing we're actually trying to do is apply a plugin to itself. However, the issue appears to be wider than this admittedly somewhat niche requirement, as I later discovered it happened no matter where I put the source files.

### Steps to Reproduce 

[This small test project](https://github.com/hakanai/self-applying-gradle-plugin) reproduces it.

Alternatively, you can do this:

1. Take some currently-working project using precompiled script plugins.
2. Rename `src/main/kotlin` to `src/main/kotlin2`
3. Add this to `build.gradle.kts`:
    ```kotlin
    kotlin {
        sourceSets["main"].kotlin.srcDir("src/main/kotlin2")
    }
    ```
    (the same happens if you use `setSrcDirs` instead.)
4. Try to build

### Your Environment

Build scan URL: https://scans.gradle.com/s/yla5icptgcms6

**Comment by hakanai** [comment]
Is the issue that this line is eagerly getting the compile classpath?

https://github.com/gradle/gradle/blob/9033c66b107f95d8bc21225c1166e496555d74a2/subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/DefaultPrecompiledScriptPluginsSupport.kt#L200

**Comment by bamboo** [comment]
Thanks for the report, @hakanai.

My hunch is that it might be related to `gradlePlugin.pluginSourceSet` not including those  plugins:

https://github.com/gradle/gradle/blob/9033c66b107f95d8bc21225c1166e496555d74a2/subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/DefaultPrecompiledScriptPluginsSupport.kt#L400

**Comment by hakanai** [comment]
So I threw in a println to see what value `gradlePlugin.pluginSourceSet` has, and it is:
```lang-none
source set 'main'
```

Which at least has the same _name_ as the source set I'm adding this source dir to...

**Comment by hakanai** [comment]
I guess `collectScriptPluginFiles` too is being called eagerly, some time before I added the source dir.

**Comment by donat** [comment]
According to the [documentation](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins)
> You write these as *.gradle files in src/main/groovy directory or *.gradle.kts files in src/main/kotlin directory.

There's no promise for precompiled script plugin source directory configurability.

**Comment by hakanai** [comment]
If you're going to take that attitude then you might as well remove `srcDirs` from the API entirely, because the docs also say things like Java sources have to be in `src/main/java`, etc. :)

**Comment by eskatos** [comment]
Agreed this can be considered a bug. The `kotlin-dsl` plugin should defer reading from the source set.
